src/sqlfluff/rules/L064.py

```diff
@@ -0,0 +1,259 @@
+                    preferred_string_quotes,
+                )
+        else:
+            preferred_string_quotes = self.preferred_string_quotes
+
+        fixed_string = self._normalize_preferred_string_quotes(
+            context.segment.raw,
+            preferred_quote_char=self._quotes_mapping[preferred_string_quotes][
+                "preferred_quote_char"
+            ],
+            alternate_quote_char=self._quotes_mapping[preferred_string_quotes][
+                "alternate_quote_char"
+            ],
+        )
+
+        if fixed_string != context.segment.raw:
+            return LintResult(
+                anchor=context.segment,
+                memory=context.memory,
+                fixes=[
+                    LintFix.replace(
+                        context.segment,
+                        [
+                            CodeSegment(
+                                raw=fixed_string,
+                                name="quoted_literal",
+                                type="literal",
+                            )
+                        ],
+                    )
+                ],
+                description=(
+                    "Inconsistent use of preferred quote style '"
+                    f"{self._quotes_mapping[preferred_string_quotes]['common_name']}'. "
+                    f"Use {fixed_string} instead of {context.segment.raw}."
+                ),
+            )
+
+        return None
+
+    def _sub_twice(self, regex: regex.Pattern, replacement: str, original: str) -> str:
+        """Replace `regex` with `replacement` twice on `original`.
+
+        This is used by string normalization to perform replaces on overlapping matches.
+        """
+        return regex.sub(replacement, regex.sub(replacement, original))
+
+    def _normalize_preferred_string_quotes(
+        self, s: str, preferred_quote_char: str, alternate_quote_char: str
+    ) -> str:
+        """Prefer `preferred_quote_char` but only if it doesn't cause more escaping.
+
+        Adds or removes backslashes as appropriate.
+        """
+        value = s.lstrip(self._string_prefix_chars)
+
+        if value[:3] == preferred_quote_char * 3:
+            # In tripple quoted strings we are not replacing escaped quotes.
+            # So nothing left to do and we can exit.
+            return s
+        elif value[0] == preferred_quote_char:
+            # Quotes are alright already. But maybe we can remove some unnecessary
+            # escapes or reduce the number of escapes using alternate_quote_char ?
+            orig_quote = preferred_quote_char
+            new_quote = alternate_quote_char
+        elif value[:3] == alternate_quote_char * 3:
+            orig_quote = alternate_quote_char * 3
+            new_quote = preferred_quote_char * 3
+        elif value[0] == alternate_quote_char:
+            orig_quote = alternate_quote_char
+            new_quote = preferred_quote_char
+        else:
+            self.logger.debug(
```

Comment:
How could this happen?

Reply:
It's a bit of a safety net. Right now, it only really does something if the rule is force enabled for dialects that allow quoted literal styles other than single or double quotes. This test is covering it: https://github.com/sqlfluff/sqlfluff/pull/3118/files#diff-ad130ac57b12a68f9bda9d3474dcb126f54009c5b3e766641d314bdd5876dbc8R231-R243

What I had in mind is: Maybe there exists (right now, or in the future) a dialect that allows single quotes, double quotes _and_ dollar quotes.